packages/protocol/src/utils/StreamPartitionID.ts

```diff
@@ -0,0 +1,154 @@
+    constructor(id: string, partition = 0) {
+        this.id = typeof id === 'string' ? id.toLowerCase() : id
+        this.partition = partition
+        this.validate()
+
+        // static values
+        const separator = (this.constructor as typeof SPID).SEPARATOR
+        this.key =`${this.id}${separator}${this.partition}`
+        this.cachedSplit = [this.id, this.partition]
+        // prevent all mutation
+        Object.freeze(this)
+        Object.freeze(this.cachedSplit)
+    }
+
+    /**
+     * Throws if data is invalid.
+     */
+    private validate() {
+        if (typeof this.id !== 'string' || this.id.length === 0) {
+            throw new Error('SPID validation failed: id must be non-empty string')
+        }
+
+        if (typeof this.partition !== 'number' || !Number.isSafeInteger(this.partition) || this.partition < 0) {
+            throw new Error('SPID validation failed: paritition must be an integer >= 0 ')
+        }
+    }
+
+    /**
+     * True iff other value is equivalent.
+     */
+    equals(other: SPIDish): boolean {
+        // check if same instance
+        if (other === this) { return true }
+        let otherSpid: SPID
+        try {
+            otherSpid = (this.constructor as typeof SPID).from(other)
+        } catch (_err) {
+            // not equal if not valid
+            return false
+        }
+
+        // check if matches both id + partition
+        return (
+            this.id === otherSpid.id
+            && this.partition === otherSpid.partition
+        )
+    }
+
+    /**
+     * Convert to SPID if possible.
+     */
+    static from(spidish: SPIDish): SPID {
+        // return spid if already spid
+        if (spidish instanceof SPID) {
+            return spidish
+        }
+
+        // convert from string
+        if (typeof spidish === 'string') {
+            return SPID.fromString(spidish)
+        } else if (spidish && typeof spidish === 'object') {
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamId = spidish.streamId || spidish.id
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamPartition = spidish.streamPartition || spidish.partition
+            return SPID.from(`${streamId}${this.SEPARATOR}${streamPartition}`)
+        } else {
+            // TODO: add more conversions?
+            throw new Error(`SPID validation failed, input is malformed: ${spidish}`)
+        }
+    }
+
+    /**
+     * Convert String to SPID if possible.
+     */
+    static fromString(spidString: string): SPID {
+        const [id, parititionStr] = spidString.split(this.SEPARATOR)
+        return new SPID(id, Number.parseFloat(parititionStr) || 0)
+    }
+
+    /**
+     * String representation of SPID id + partition
+     */
+    toString(): string {
+        return this.key
+    }
+
+    /**
+     * Alias of toString.
+     */
+    toKey(): string {
+        return this.key
```

Comment:
Just to be safe (and faithful to the code comment) in the face of potential future modifications, I would have the implementation here call `this.toString()`.

Reply:
Right, sure